api/admin/controller.py

```diff
@@ -291,87 +305,70 @@ class SignInController(AdminController):
 </body>
 </html>"""
 
-    PASSWORD_SIGN_IN_TEMPLATE = """<!DOCTYPE HTML>
+    SIGN_IN_TEMPLATE = """<!DOCTYPE HTML>
 <html lang="en">
 <head><meta charset="utf8"></head>
-</body>
-<form action="%(password_sign_in_url)s" method="post">
-<input type="hidden" name="redirect" value="%(redirect)s"/>
-<label>Email <input type="text" name="email" /></label>
-<label>Password <input type="password" name="password" /></label>
-<button type="submit">Sign In</button>
-</form>
+<body>
+%(auth_provider_html)s
 </body>
 </html>"""
 
 
     def sign_in(self):
-        """Redirects admin if they're signed in."""
-        if not self.auth:
+        """Redirects admin if they're signed in, or shows the sign in page."""
+        if not self.auth_providers:
             return ADMIN_AUTH_NOT_CONFIGURED
 
         admin = self.authenticated_admin_from_request()
 
         if isinstance(admin, ProblemDetail):
             redirect_url = flask.request.args.get("redirect")
-            return redirect(self.auth.auth_uri(redirect_url), Response=Response)
+            auth_provider_html = [auth.sign_in_template(redirect_url) for auth in self.auth_providers]
+            auth_provider_html = "<br/><hr/>or<br/><br/>".join(auth_provider_html)
+            html = self.SIGN_IN_TEMPLATE % dict(
+                auth_provider_html=auth_provider_html
+            )
+            headers = dict()
+            headers['Content-Type'] = "text/html"
+            return Response(html, 200, headers)
         elif admin:
             return redirect(flask.request.args.get("redirect"), Response=Response)
 
     def redirect_after_google_sign_in(self):
         """Uses the Google OAuth client to determine admin details upon
         callback. Barring error, redirects to the provided redirect url.."""
-        if not self.auth:
+        if not self.auth_providers:
             return ADMIN_AUTH_NOT_CONFIGURED
 
-        if not isinstance(self.auth, GoogleOAuthAdminAuthenticationProvider):
+        auth = self.auth_provider(GoogleOAuthAdminAuthenticationProvider.NAME)
+        if not auth:
             return ADMIN_AUTH_MECHANISM_NOT_CONFIGURED
 
-        admin_details, redirect_url = self.auth.callback(flask.request.args)
+        admin_details, redirect_url = auth.callback(flask.request.args)
         if isinstance(admin_details, ProblemDetail):
             return self.error_response(admin_details)
 
-        if not self.staff_email(admin_details['email']):
+        if not auth.staff_email(self._db, admin_details['email']):
```

Comment:
Since you're moving `staff_email` into the Authenticator class you might have an opportunity to streamline this method by having `callback` also check `staff_email` and return INVALID_ADMIN_CREDENTIALS if there's a problem.